Adding error page and future vision page to Neon Authorize (#2394)
neondatabase/website: content/docs/guides/neon-authorize-future.md

---
title: The future for Neon Authorize
subtitle: Our vision for what comes next
enableTableOfContents: true
---

> [Neon Authorize](/docs/guides/neon-authorize) is an exciting feature, and we want to build more on top of it.

Reviewer: is an exciting feature, and we want to build more capabilities on top of it.

> Here's what we'd like to work on next — in no particular order (unless stated otherwise). If you see something you like, or something you'd like to see but don't, let us know in [Discord](https://discord.com/channels/1176467419317940276/1176788564890112042)!

Reviewer: Here are some ideas of things that we could build next – in no particular order...

Author: fixed

## Accepting JWTs for TCP and WebSocket connections

> We want Neon Authorize to work over any type of connection: TCP, HTTP, or WebSockets. This is probably our highest priority. Accepting JWTs over TCP (in the password field for the Postgres role) would mean that any Postgres SDK in any programming language could support Neon Authorize. We will build this.

Reviewer: Remove "We will build this" – it's not commited

## Custom roles

> We aim to introduce custom roles for Neon Authorize, beyond the "authenticated" and "anonymous" roles we have now. We're also exploring the idea of adding "Neon-managed", which would be managed by our platform and included by default in every branch and read replica.

Reviewer: "Neon-managed" roles

Author: fixed

> We aim to introduce custom roles for Neon Authorize, beyond the "authenticated" and "anonymous" roles we have now. We're also exploring the idea of adding "Neon-managed", which would be managed by our platform and included by default in every branch and read replica.

Reviewer 2: ```suggestion
We aim to introduce custom roles for Neon Authorize, in addition to the "authenticated" and "anonymous" roles we have now. We're also exploring the idea of adding a "Neon-managed" role, which would be managed by our platform and included by default in every branch and read replica.
```

Reviewer 2: @davidgomes I am wondering if users are interested in us wanting to add a Neon-managed role. Perhaps I'm missing the value/purpose of this role from an end user's perspective.

Reviewer: The idea behind Neon-managed roles is that we can guarantee they'll always exist (i.e., they can never be dropped). This is very useful for Neon Authorize in specific, because you can then guarantee that `authenticated` and `anonymous` will always exist. As it stands today, someone could accidentally delete these roles.

## JWT audience checks

The `aud` field in a JWT typically defines the token's intended recipients. In the future, we want to allow users to customize the expected `aud` field in the JWTs they use with Neon Authorize, providing greater flexibility and control.

## Multi-tenancy through Neon Authorize

> We want to support multi-tenancy with Neon Authorize, allowing users to configure JWTs that trigger Neon to route database requests to different Neon instances.

Reviewer: "We've discussed supporting multi-tenancy via Neon Authorize, allowing users to configure how their JWTs should trigger Neon to route database requests to different Neon instances"

## Neon CLI for configuration

> We can add the ability to configure Neon Authorize from our Neon CLI.

Reviewer 2: ```suggestion
We would like to add support for configuring Neon Authorize from the Neon CLI.
```

## Neon Authorize for Neon Console frontend database requests

> Using Neon Authorize for database requests from our SQL Editor can help map the individual Neon Console user more easily to the person in the organization who triggered the request.

Reviewer 2: ```suggestion
Using Neon Authorize for database requests from the Neon SQL Editor can help map the Neon Console user to the individual who issued the request.
```

## Proxy rate limiting

> Implementing rate limiting will make SQL from the client safer.

Reviewer 2: ```suggestion
Implementing proxy rate limiting will make SQL from the client safer.
```

## Query allow-listing

> A query allow-list would also help increase the safety of SQL queries from the client.

Reviewer 2: ```suggestion
The ability to define a query allow-list would help increase the safety of SQL queries from the client.
```

## Investigate React/TanStack query hooks for SQL queries

> We're exploring the generation of React Query hooks for SQL queries, simplifying the process of building applications that leverage SQL from the front end.

Reviewer: that leverage SQL from the frontend, or that use React Server Components.

(frontend should be one word I think)

Author: fixed

## Simplified OAuth flow for JWKS endpoint

> Instead of manually copying and pasting the JWKS endpoint from your authentication provider's console, we're working on an OAuth flow that allows you to simply `Sign in with <Provider>`, automatically populating the JWKS endpoint.

Reviewer: "we could work on an OAuth flow"

again, nothing on this list is commited so it's best if we stay defensive

Author: fixed

> Instead of manually copying and pasting the JWKS endpoint from your authentication provider's console, we're working on an OAuth flow that allows you to simply `Sign in with <Provider>`, automatically populating the JWKS endpoint.

Reviewer 2: ```suggestion
Instead of manually copying and pasting the JWKS endpoint from your authentication provider's console, we are interested in developing an OAuth flow that allows you to simply `Sign in with <Provider>`, automatically populating the JWKS endpoint.
```

## User interface for RLS and user impersonation

> We'd like to add a friendly UI for managing RLS policies, with AI assistance plus a way to simulate SQL queries from different application users and JWT properties.

Reviewer 2: ```suggestion
We'd like to add a user-friendly UI for managing RLS policies, with AI assistance, and a way to simulate SQL queries from different application users and JWT properties.
```
